### PR TITLE
Improve x-limit, y-limit, and rasterize support for geo features 

### LINKED ISSUE #1090: Tile + multiple Point breaks
#### ALL software version info

- geoviews 1.10.0
- hvplot 0.8.4
- holoviews 1.16.2

#### Description of expected behavior and the observed behavior

Overlapping a tile plus multiple Points plots breaks.
This used to be working with earlier versions.

#### Complete, minimal, self-contained example code that reproduces the issue

```
import pandas as pd
import hvplot.pandas

df = pd.DataFrame(dict(lon=[0, 10], lat=[40, 50], v=[0,1]))

(
df.hvplot.points(x="lon", y="lat", s=200, c="y", geo=True, tiles='CartoLight')
*df.hvplot.points(x="lon", y="lat", c="v", geo=True)
)
```

#### Stack traceback and/or browser JavaScript console output

```
---------------------------------------------------------------------------
Exception                                 Traceback (most recent call last)
File ~/.miniconda3/envs/pynsitu/lib/python3.10/site-packages/IPython/core/formatters.py:974, in MimeBundleFormatter.__call__(self, obj, include, exclude)
    971     method = get_real_method(obj, self.print_method)
    973     if method is not None:
--> 974         return method(include=include, exclude=exclude)
    975     return None
    976 else:

File ~/.miniconda3/envs/pynsitu/lib/python3.10/site-packages/holoviews/core/dimension.py:1290, in Dimensioned._repr_mimebundle_(self, include, exclude)
   1283 def _repr_mimebundle_(self, include=None, exclude=None):
   1284     """
   1285     Resolves the class hierarchy for the class rendering the
   1286     object using any display hooks registered on Store.display
   1287     hooks.  The output of all registered display_hooks is then
   1288     combined and returned.
   1289     """
-> 1290     return Store.render(self)

File ~/.miniconda3/envs/pynsitu/lib/python3.10/site-packages/holoviews/core/options.py:1425, in Store.render(cls, obj)
   1423 data, metadata = {}, {}
   1424 for hook in hooks:
-> 1425     ret = hook(obj)
   1426     if ret is None:
   1427         continue

File ~/.miniconda3/envs/pynsitu/lib/python3.10/site-packages/holoviews/ipython/display_hooks.py:280, in pprint_display(obj)
    278 if not ip.display_formatter.formatters['text/plain'].pprint:
    279     return None
--> 280 return display(obj, raw_output=True)

File ~/.miniconda3/envs/pynsitu/lib/python3.10/site-packages/holoviews/ipython/display_hooks.py:248, in display(obj, raw_output, **kwargs)
    246 elif isinstance(obj, (CompositeOverlay, ViewableElement)):
    247     with option_state(obj):
--> 248         output = element_display(obj)
    249 elif isinstance(obj, (Layout, NdLayout, AdjointLayout)):
    250     with option_state(obj):

File ~/.miniconda3/envs/pynsitu/lib/python3.10/site-packages/holoviews/ipython/display_hooks.py:142, in display_hook.<locals>.wrapped(element)
    140 try:
    141     max_frames = OutputSettings.options['max_frames']
--> 142     mimebundle = fn(element, max_frames=max_frames)
    143     if mimebundle is None:
    144         return {}, {}

File ~/.miniconda3/envs/pynsitu/lib/python3.10/site-packages/holoviews/ipython/display_hooks.py:188, in element_display(element, max_frames)
    185 if type(element) not in Store.registry[backend]:
    186     return None
--> 188 return render(element)

File ~/.miniconda3/envs/pynsitu/lib/python3.10/site-packages/holoviews/ipython/display_hooks.py:69, in render(obj, **kwargs)
     66 if renderer.fig == 'pdf':
     67     renderer = renderer.instance(fig='png')
---> 69 return renderer.components(obj, **kwargs)

File ~/.miniconda3/envs/pynsitu/lib/python3.10/site-packages/holoviews/plotting/renderer.py:399, in Renderer.components(self, obj, fmt, comm, **kwargs)
    396 embed = (not (dynamic or streams or self.widget_mode == 'live') or config.embed)
    398 if embed or config.comms == 'default':
--> 399     return self._render_panel(plot, embed, comm)
    400 return self._render_ipywidget(plot)

File ~/.miniconda3/envs/pynsitu/lib/python3.10/site-packages/holoviews/plotting/renderer.py:406, in Renderer._render_panel(self, plot, embed, comm)
    404 doc = Document()
    405 with config.set(embed=embed):
--> 406     model = plot.layout._render_model(doc, comm)
    407 if embed:
    408     return render_model(model, comm)

File ~/.miniconda3/envs/pynsitu/lib/python3.10/site-packages/panel/viewable.py:736, in Viewable._render_model(self, doc, comm)
    734 if comm is None:
    735     comm = state._comm_manager.get_server_comm()
--> 736 model = self.get_root(doc, comm)
    738 if self._design and self._design.theme.bokeh_theme:
    739     doc.theme = self._design.theme.bokeh_theme

File ~/.miniconda3/envs/pynsitu/lib/python3.10/site-packages/panel/layout/base.py:286, in Panel.get_root(self, doc, comm, preprocess)
    282 def get_root(
    283     self, doc: Optional[Document] = None, comm: Optional[Comm] = None,
    284     preprocess: bool = True
    285 ) -> Model:
--> 286     root = super().get_root(doc, comm, preprocess)
    287     # ALERT: Find a better way to handle this
    288     if hasattr(root, 'styles') and 'overflow-x' in root.styles:

File ~/.miniconda3/envs/pynsitu/lib/python3.10/site-packages/panel/viewable.py:658, in Renderable.get_root(self, doc, comm, preprocess)
    656 wrapper = self._design._wrapper(self)
    657 if wrapper is self:
--> 658     root = self._get_model(doc, comm=comm)
    659     if preprocess:
    660         self._preprocess(root)

File ~/.miniconda3/envs/pynsitu/lib/python3.10/site-packages/panel/layout/base.py:170, in Panel._get_model(self, doc, root, parent, comm)
    168 root = root or model
    169 self._models[root.ref['id']] = (model, parent)
--> 170 objects, _ = self._get_objects(model, [], doc, root, comm)
    171 props = self._get_properties(doc)
    172 props[self._property_mapping['objects']] = objects

File ~/.miniconda3/envs/pynsitu/lib/python3.10/site-packages/panel/layout/base.py:155, in Panel._get_objects(self, model, old_objects, doc, root, comm)
    153 else:
    154     try:
--> 155         child = pane._get_model(doc, root, model, comm)
    156     except RerenderError:
    157         return self._get_objects(model, current_objects[:i], doc, root, comm)

File ~/.miniconda3/envs/pynsitu/lib/python3.10/site-packages/panel/pane/holoviews.py:380, in HoloViews._get_model(self, doc, root, parent, comm)
    378     plot = self.object
    379 else:
--> 380     plot = self._render(doc, comm, root)
    382 plot.pane = self
    383 backend = plot.renderer.backend

File ~/.miniconda3/envs/pynsitu/lib/python3.10/site-packages/panel/pane/holoviews.py:461, in HoloViews._render(self, doc, comm, root)
    458     if comm:
    459         kwargs['comm'] = comm
--> 461 return renderer.get_plot(self.object, **kwargs)

File ~/.miniconda3/envs/pynsitu/lib/python3.10/site-packages/holoviews/plotting/bokeh/renderer.py:70, in BokehRenderer.get_plot(self_or_cls, obj, doc, renderer, **kwargs)
     63 @bothmethod
     64 def get_plot(self_or_cls, obj, doc=None, renderer=None, **kwargs):
     65     """
     66     Given a HoloViews Viewable return a corresponding plot instance.
     67     Allows supplying a document attach the plot to, useful when
     68     combining the bokeh model with another plot.
     69     """
---> 70     plot = super().get_plot(obj, doc, renderer, **kwargs)
     71     if plot.document is None:
     72         plot.document = Document() if self_or_cls.notebook_context else curdoc()

File ~/.miniconda3/envs/pynsitu/lib/python3.10/site-packages/holoviews/plotting/renderer.py:236, in Renderer.get_plot(self_or_cls, obj, doc, renderer, comm, **kwargs)
    234 if isinstance(obj, AdjointLayout):
    235     obj = Layout(obj)
--> 236 plot = self_or_cls.plotting_class(obj)(obj, renderer=renderer,
    237                                        **plot_opts)
    238 defaults = [kd.default for kd in plot.dimensions]
    239 init_key = tuple(v if d is None else d for v, d in
    240                  zip(plot.keys[0], defaults))

File ~/.miniconda3/envs/pynsitu/lib/python3.10/site-packages/geoviews/plotting/bokeh/plot.py:179, in GeoOverlayPlot.__init__(self, element, **params)
    178 def __init__(self, element, **params):
--> 179     super().__init__(element, **params)
    180     self.geographic = any(element.traverse(is_geographic, [_Element]))
    181     if self.geographic:

File ~/.miniconda3/envs/pynsitu/lib/python3.10/site-packages/geoviews/plotting/bokeh/plot.py:57, in GeoPlot.__init__(self, element, **params)
     56 def __init__(self, element, **params):
---> 57     super().__init__(element, **params)
     58     self.geographic = is_geographic(self.hmap.last)
     59     if self.geographic and not isinstance(self.projection, (PlateCarree, Mercator)):

File ~/.miniconda3/envs/pynsitu/lib/python3.10/site-packages/holoviews/plotting/plot.py:1535, in GenericOverlayPlot.__init__(self, overlay, ranges, batched, keys, group_counter, **params)
   1533 def __init__(self, overlay, ranges=None, batched=True, keys=None, group_counter=None, **params):
   1534     if 'projection' not in params:
-> 1535         params['projection'] = self._get_projection(overlay)
   1537     super().__init__(overlay, ranges=ranges, keys=keys,
   1538                      batched=batched, **params)
   1540     # Apply data collapse

File ~/.miniconda3/envs/pynsitu/lib/python3.10/site-packages/geoviews/plotting/plot.py:42, in ProjectionPlot._get_projection(self, obj)
     39 custom_projs = [p for p in projections if p is not None]
     41 if len(set([type(p) for p in custom_projs])) > 1:
---> 42     raise Exception("An axis may only be assigned one projection type")
     43 elif custom_projs:
     44     return custom_projs[0]

Exception: An axis may only be assigned one projection type
```

### LINKED ISSUE #1093: Vectorfield not overlaid properly on Tiles
https://hvplot.holoviz.org/reference/xarray/vectorfield.html#geographic-data
image

### PR DESCRIPTION
**Pull request by Hoxbro**
Fixes #1093 
Fixes #1090
Fixes  https://discourse.holoviz.org/t/any-smart-approach-for-determining-what-holoviz-package-broke-my-notebook/5944 

If geo is true, we will use GeoViews tile sources, and if false, we will use HoloViews. GeoViews features and tiles are projected to the respective projection. 

---
1093 code:
``` python
import cartopy.crs as ccrs
import geoviews as gv
import hvplot.xarray  # noqa
import numpy as np
import xarray as xr


def sample_data(shape=(20, 30)):
    """
    Return ``(x, y, u, v, crs)`` of some vector data
    computed mathematically. The returned crs will be a rotated
    pole CRS, meaning that the vectors will be unevenly spaced in
    regular PlateCarree space.

    """
    crs = ccrs.RotatedPole(pole_longitude=177.5, pole_latitude=37.5)

    x = np.linspace(311.9, 391.1, shape[1])
    y = np.linspace(-23.6, 24.8, shape[0])

    x2d, y2d = np.meshgrid(x, y)
    u = 10 * (2 * np.cos(2 * np.deg2rad(x2d) + 3 * np.deg2rad(y2d + 30)) ** 2)
    v = 20 * np.cos(6 * np.deg2rad(x2d))

    return x, y, u, v, crs


xs, ys, U, V, crs = sample_data()

mag = np.sqrt(U**2 + V**2)
angle = (np.pi / 2.0) - np.arctan2(U / mag, V / mag)

ds = xr.Dataset(
    {
        "mag": xr.DataArray(mag, dims=("y", "x"), coords={"y": ys, "x": xs}),
        "angle": xr.DataArray(angle, dims=("y", "x"), coords={"y": ys, "x": xs}),
    },
    attrs={"crs": crs},
)
ds.hvplot.vectorfield(
    x="x",
    y="y",
    angle="angle",
    mag="mag",
    hover=False,
    geo=True,
    tiles="CartoLight",
)
```
image

1090 code:
``` python
import hvplot.pandas
import pandas as pd

df = pd.DataFrame({"lon": [0, 10], "lat": [40, 50], "v": [0, 1]})

(
    df.hvplot.points(x="lon", y="lat", s=200, c="y", geo=True, tiles="CartoLight")
    * df.hvplot.points(x="lon", y="lat", c="v", geo=True)
)
```
image

---
Code from #1053 (Comments are from original PR; it seems that every case works)
``` python
import xarray as xr
import hvplot.xarray

dz = xr.tutorial.open_dataset('air_temperature')
dz.hvplot('lon', 'lat', coastline=True).opts(ylim=(20, 30), xlim=(-160, -140))
````
image

Case 1:
``` python
import hvplot.pandas  # noqa
import cartopy.crs as ccrs
import geoviews as gv
from bokeh.sampledata.airport_routes import airports

# Case 1: This example works
airports.hvplot.points(
    "Longitude",
    "Latitude",
    color="red",
    alpha=0.2,
    xlim=(-180, -30),
    ylim=(0, 72),
    geo=True,
    tiles="OSM",
)
```
image

``` python
# Case 2: xlim/ylim are not obeyed when the map tile is removed
airports.hvplot.points(
    "Longitude",
    "Latitude",
    geo=True,
    color="red",
    alpha=0.2,
    xlim=(-180, -30),
    ylim=(0, 72),
)
```
image

``` python
# Case 3: Setting ymin > 0 raises errors
# Change ymin from 0 to 10
airports.hvplot.points(
    "Longitude",
    "Latitude",
    geo=True,
    color="red",
    alpha=0.2,
    xlim=(-180, -30),
    ylim=(10, 72),
)
````
image

``` python
# Case 4: Setting ymax < 0 also raises errors
airports.hvplot.points(
    "Longitude",
    "Latitude",
    geo=True,
    color="red",
    alpha=0.2,
    xlim=(-180, -30),
    ylim=(-20, -5),
)
```
It works but gives a long empty plot, with one red dot.

``` python
# Works if geo=False
airports.hvplot.points(
    "Longitude",
    "Latitude",
    geo=False,
    color="red",
    alpha=0.2,
    xlim=(-180, -30),
    ylim=(10, 72),
)
```

image

### CODE REVIEW

File: hvplot/converter.py [hunk 1]
```diff
@@ -1411,32 +1410,44 @@ def _apply_layers(self, obj):
...
+                self.tiles,
+                gv.tile_sources.tile_sources,
+                (gv.element.WMTS, hv.element.tiles.Tiles),
+            )
             obj = tiles * obj
         return obj
 
+    def _get_tiles(self, source, sources, types):
+        tile_source = 'EsriImagery' if self.tiles == 'ESRI' else self.tiles
+        warning = f"{tile_source} tiles not recognized, must be one of: {sorted(sources)} or a tile object"
```
Comment (ahuang11) [inline hvplot/converter.py:1438]: I don't think this should be called warning. Also I think this should be put directly in the else clause for clarity.

**Comment by ahuang11** [comment]
This looks good. I just wanted to ask whether this should this work (xlim/ylim in hvplot call), or out of scope in this PR?

```python
import hvplot.xarray
import xarray as xr
import cartopy.crs as ccrs

ds = xr.tutorial.open_dataset("air_temperature")
ds.hvplot.points(
    "lon", "lat", crs=ccrs.PlateCarree(), rasterize=True, features=["coastline", "land"],
    xlim=(-140, -80), ylim=(20, 45)
)
```

This works:

```python
import hvplot.xarray
import xarray as xr
import cartopy.crs as ccrs

ds = xr.tutorial.open_dataset("air_temperature")
ds.hvplot.points(
    "lon", "lat", crs=ccrs.PlateCarree(), rasterize=True, features=["coastline", "land"],
).opts(xlim=(-140, -80), ylim=(20, 45))
```

**Comment by ahuang11** [comment]
Okay, confusingly:

This doesn't work (xlim, ylim in opts with tiles=True)
```python
import hvplot.xarray
import xarray as xr
import cartopy.crs as ccrs

ds = xr.tutorial.open_dataset("air_temperature")
ds.hvplot.points(
    "lon", "lat", crs=ccrs.PlateCarree(), tiles=True
).opts(xlim=(-140, -80), ylim=(20, 45))
```
image


But this works:
```python
import hvplot.xarray
import xarray as xr
import cartopy.crs as ccrs

ds = xr.tutorial.open_dataset("air_temperature")
ds.hvplot.points(
    "lon", "lat", crs=ccrs.PlateCarree(), tiles=True, xlim=(-140, -80), ylim=(20, 45)
)
```
image

File: hvplot/converter.py [hunk 2]
```diff
@@ -446,20 +446,19 @@ def __init__(
                         "Projection must be defined as cartopy CRS or "
                         f"one of the following CRS string:\n {all_crs}")
 
-            projection = projection or (ccrs.GOOGLE_MERCATOR if tiles else self.crs)
-            if tiles and projection != ccrs.GOOGLE_MERCATOR:
+            self.proj_crs = projection or (ccrs.GOOGLE_MERCATOR if tiles else self.crs)
+            if tiles and self.proj_crs != ccrs.GOOGLE_MERCATOR:
```
Comment (ahuang11) [inline hvplot/converter.py:450]: To help future developers understand, should we rename `proj_crs` to `feature_projection`?

**Comment by ahuang11** [comment]
Not sure if this is out of scope of this PR, but this isn't correct: (with rasterize=True + projection)
```python
import hvplot.xarray
import xarray as xr
import cartopy.crs as ccrs

ds = xr.tutorial.open_dataset("air_temperature")
ds.hvplot.points(
    "lon", "lat", crs=ccrs.PlateCarree(), features=["coastline", "land"],
    xlim=(-140, -80), ylim=(20, 45), projection=ccrs.LambertConformal(), rasterize=True
)
```

image

Dropping rasterize=True works:

```python
import hvplot.xarray
import xarray as xr
import cartopy.crs as ccrs

ds = xr.tutorial.open_dataset("air_temperature")
ds.hvplot.points(
    "lon", "lat", crs=ccrs.PlateCarree(), features=["coastline", "land"],
    projection=ccrs.LambertConformal()
)
```
image

**Comment by Hoxbro** [comment]
`rasterize` and `datashade` with geographic operations are out of scope for this PR. Though I will give it a try, but I won't spend too long on it. 

At some point, I want it only to be needed to pass limits directly into hvplot like `hvplot(..., xlim=(...) ylim=(...)`.

**Comment by Hoxbro** [comment]
@ahuang11, can you try to play around with the latest commit?
``` python
import hvplot.xarray
import xarray as xr
import cartopy.crs as ccrs

ds = xr.tutorial.open_dataset("air_temperature")
ds.hvplot.points(
    "lon",
    "lat",
    crs=ccrs.PlateCarree(),
    features=["coastline", "land"],
    xlim=(-140, -80),
    ylim=(20, 45),
    projection=ccrs.LambertConformal(),
    rasterize=True,
    dynspread=True,
)
```

image

**Comment by ahuang11** [comment]
Thanks! Most of the things I tried worked. This still doesn't work:
```python
import hvplot.xarray
import xarray as xr
import cartopy.crs as ccrs

ds = xr.tutorial.open_dataset("air_temperature")
ds.hvplot.points(
    "lon", "lat", crs=ccrs.PlateCarree(), projection=ccrs.GOOGLE_MERCATOR, tiles=True
).opts(xlim=(-140, -80), ylim=(20, 45))
```
Dropping tiles works.

**Comment by Hoxbro** [comment]
But putting `xlim` and `ylim` inside the hvplot call works?

**Comment by ahuang11** [comment]
Yes

**Comment by Hoxbro** [comment]
I think this is more of a problem with GeoViews with projected overlays as this will also zoom in to small numbers:

``` python
import cartopy.crs as ccrs
import geoviews as gv

ds = xr.tutorial.open_dataset("air_temperature")
plot = gv.Points(ds.isel(time=0), kdims=["lon", "lat"]).opts(projection=ccrs.GOOGLE_MERCATOR)
(plot * plot).opts(xlim=(-140, -80), ylim=(20, 45))
```
image

**Comment by ahuang11** [comment]
Right, I'm not sure how it gets handled without overlays, but xlim/ylim is not being projected properly in opts when there's an overlay.

**Comment by Hoxbro** [comment]
My example above works fine if there is no overlaid: `plot.opts(xlim=(-140, -80), ylim=(20, 45))`.

I don't think we can do anything on the hvplot side of things.

**Comment by ahuang11** [comment]
It might be related to this:
https://github.com/holoviz/geoviews/issues/340

**Comment by ahuang11** [comment]
One last comment is can we rename `proj_crs` to be more descriptive (like feature_projection)?

File: hvplot/converter.py [hunk 3]
```diff
@@ -1363,6 +1362,10 @@ def method_wrapper(ds, x, y):
             if self._dim_ranges.get('c', (None, None)) != (None, None):
                 style['clim'] = self._dim_ranges['c']
 
+        if self.geo:
+            import geoviews as gv
+            obj = gv.project(obj, projection=self.proj_crs)
```
Comment (ahuang11) [inline hvplot/converter.py:1367]: Wondering if this should be after processed = operation() because I worry this eagerly loads the entire dataset.

**Comment by Hoxbro** [comment]
I don't mind a better name, but I'm not sure if it's correct to call it `feature_projection`. It is used for the features but also for other calculations like finding limits and the geo projection.

**Comment by ahuang11** [comment]
maybe `output_projection`?

File: hvplot/converter.py (hunk 3 above)
Comment (Hoxbro) [reply hvplot/converter.py:1367]: It is needed before the operation. If you move it after it, you get the following figure. We could add more checks to the if statement so we only do this when it is absolutely needed. 

image

``` python
import hvplot.xarray
import xarray as xr
import cartopy.crs as ccrs

ds = xr.tutorial.open_dataset("air_temperature")
ds.hvplot.points(
    "lon",
    "lat",
    crs=ccrs.PlateCarree(),
    features=["coastline", "land"],
    xlim=(-140, -80),
    ylim=(20, 45),
    projection=ccrs.LambertConformal(),
    rasterize=True,
    dynspread=True,
)
```

Comment (ahuang11) [reply hvplot/converter.py]: Yes I support more checks. I imagine someone has a huge dataset, and gv.project causes it to eagerly load, causing their CPU to crash. Not sure if this should be implemented in another PR, or what the best way forward is. 

Without this change, datashade/rasterize is effectively broken with projections?

**Comment by ahuang11** [comment]
So I'm using the latest released version of hvplot:

I wonder how image is handled under the hood (i.e. why does this work)
```python
import hvplot.xarray
import xarray as xr
import cartopy.crs as ccrs

ds = xr.tutorial.open_dataset("air_temperature")
ds.hvplot.image(
    "lon", "lat", crs=ccrs.PlateCarree(), features=["coastline", "land"],
    projection=ccrs.LambertConformal(), rasterize=True
)
```

vs points (but not this?):
```python
import hvplot.xarray
import xarray as xr
import cartopy.crs as ccrs

ds = xr.tutorial.open_dataset("air_temperature")
ds.hvplot.points(
    "lon", "lat", crs=ccrs.PlateCarree(), features=["coastline", "land"],
    projection=ccrs.LambertConformal(), rasterize=True
)
```

File: hvplot/converter.py (hunk 3 above)
Comment (Hoxbro) [reply hvplot/converter.py]: I have added one more check. I don't think a HoloViews operation will load data more eagerly than another. Both `gv.project` and `rasterize` are operations, which should work with different data formats. 

By the look of it, it does not work great.

**Comment by Hoxbro** [comment]
The latest release of hvplot has problems with xlimit and ylimit, which is what this PR is trying to fix.

**Comment by ahuang11** [comment]
All in all, I think this PR generally improves geo support!

Do you think a test should be added for rasterize + points?

**Comment by Hoxbro** [comment]
> Do you think a test should be added for rasterize + points?

Yes! But I didn't want to do it before we agreed this was the right approach.

**Comment by ahuang11** [comment]
> The latest release of hvplot has problems with xlimit and ylimit, which is what this PR is trying to fix.

Right; I suppose I should have commented under `obj = gv.project(obj, projection=self.proj_crs)`; my question is why does a rasterized image project properly but not rasterized points.

**Comment by Hoxbro** [comment]
> why does a rasterized image project properly but not rasterized points

I don't know. But if I move around the image on the latest release it disappears, which does not happen after this PR.

**Comment by ahuang11** [comment]
Okay; I think after tests are added, this can be merged.

**Review by ahuang11 — APPROVED** [review]

**Pull request merged**